Groundwork for oximeter instance vCPU metrics

bnaecker opened the pull request:
- Adds the silo and project IDs to the instance-ensure request from Nexus to the sled-agent. These are used as fields on the instance-related statistics. This metadata is currently unused, but will be forwarded to Propolis in the instance-ensure request once the server is updated to accept it.
- Defines a `VirtualMachine` oximeter target and `VcpuUsage` metric. The latter has a `state` field which corresponds to the named kstats published by the hypervisor that accumulate the time spent in a number of vCPU microstates. The combination of these should allow us to aggregate or break down vCPU usage by silo, project, instance, vCPU ID, and CPU state. Adds some simple mocks and tests for these.
- Adds more fine-grained feature flags to the `oximeter-instruments` crate.

bnaecker commented:
I think this ready for some initial reviews. I've tested everything several times using the Propolis branch [here](https://github.com/oxidecomputer/propolis/tree/vcpu-usage-stats). I ran the following components:

- ClickHouse server directly, using `omicron-dev ch-run`
- `oximeter` in standalone mode, with:  `cargo r --bin oximeter -- standalone --clickhouse '[::1]:8123' --log-level debug`
- `proplis-server` running a VM following the instructions [here](https://github.com/oxidecomputer/propolis/tree/vcpu-usage-stats/bin/propolis-server#running): `pfexec ./target/release/propolis-server run metrics-test.toml --metric-addr '[::1]:12221' '[::1]:12400'`


Propolis registers with the mocked control plane components, and starts producing data right away (which is all zeros, since the VM hasn't started). Here's a snippet of the Propolis server logs:

```
Jan 26 17:34:19.150 INFO starting oximeter metric producer server, interval: 30s, address: [::1]:40818, producer_id: 02d2f58c-43aa-4ebf-a07b-113c0a446716, route: /collect/02d2f58c-43aa-4ebf-a07b-113c0a446716, file: /home/bnaecker/.cargo/git/checkouts/omicron-d039c41f152bda83/f27b112/oximeter/producer/src/lib.rs:203, component: oximeter-producer, uri: /instance, method: PUT, req_id: d6c5ee8a-aadd-47fe-96f3-3848e84752ec, remote_addr: [::1]:53099, local_addr: [::1]:12400
Jan 26 17:34:19.150 INFO registered as a metric producer with Nexus, uri: /instance, method: PUT, req_id: d6c5ee8a-aadd-47fe-96f3-3848e84752ec, remote_addr: [::1]:53099, local_addr: [::1]:12400
Jan 26 17:34:49.182 INFO accepted connection, remote_addr: [::1]:37675, local_addr: [::1]:40818, component: dropshot, file: /home/bnaecker/.cargo/git/checkouts/dropshot-a4a923d29dccc492/605a6d1/dropshot/src/server.rs:765, component: oximeter-producer, uri: /instance, method: PUT, req_id: d6c5ee8a-aadd-47fe-96f3-3848e84752ec, remote_addr: [::1]:53099, local_addr: [::1]:12400
Jan 26 17:34:49.183 INFO request completed, latency_us: 1126, response_code: 200, uri: /collect/02d2f58c-43aa-4ebf-a07b-113c0a446716, method: GET, req_id: dc6f69fd-8b77-4e2c-b393-d9afef1f3688, remote_addr: [::1]:37675, local_addr: [::1]:40818, component: dropshot, file: /home/bnaecker/.cargo/git/checkouts/dropshot-a4a923d29dccc492/605a6d1/dropshot/src/server.rs:849, component: oximeter-producer, uri: /instance, method: PUT, req_id: d6c5ee8a-aadd-47fe-96f3-3848e84752ec, remote_addr: [::1]:53099, local_addr: [::1]:12400
```

That shows registration and first collection from `oximeter`. We can see `oximeter` successfully collecting and inserting the samples here:

```
Jan 26 17:34:53.691 DEBG new timeseries schema, schema: {"timeseries_name":"virtual_machine:vcpu_usage","fields.name":["instance_id","project_id","silo_id","state","vcpu_id"],"fields.type":["Uuid","Uuid","Uuid","String","U32"],"fields.source":["Target","Target","Target","Metric","Metric"],"datum_type":"CumulativeU64","created":"2024-01-26 17:34:53.688998673"}, timeseries_name: virtual_machine:vcpu_usage, id: 0b8beb02-0822-4db0-b5b7-279e06493cc9, component: clickhouse-client, collector_id: 427f0ed7-519f-4fc9-bbf6-df18557bcddb, component: oximeter-standalone, component: nexus-standalone
Jan 26 17:34:53.745 DEBG inserting 4 new timeseries schema, id: 0b8beb02-0822-4db0-b5b7-279e06493cc9, component: clickhouse-client, collector_id: 427f0ed7-519f-4fc9-bbf6-df18557bcddb, component: oximeter-standalone, component: nexus-standalone
Jan 26 17:34:53.749 DEBG inserted rows into table, table_name: oximeter.fields_string, n_rows: 24, id: 0b8beb02-0822-4db0-b5b7-279e06493cc9, component: clickhouse-client, collector_id: 427f0ed7-519f-4fc9-bbf6-df18557bcddb, component: oximeter-standalone, component: nexus-standalone
Jan 26 17:34:53.751 DEBG inserted rows into table, table_name: oximeter.fields_u32, n_rows: 24, id: 0b8beb02-0822-4db0-b5b7-279e06493cc9, component: clickhouse-client, collector_id: 427f0ed7-519f-4fc9-bbf6-df18557bcddb, component: oximeter-standalone, component: nexus-standalone
Jan 26 17:34:53.754 DEBG inserted rows into table, table_name: oximeter.fields_uuid, n_rows: 81, id: 0b8beb02-0822-4db0-b5b7-279e06493cc9, component: clickhouse-client, collector_id: 427f0ed7-519f-4fc9-bbf6-df18557bcddb, component: oximeter-standalone, component: nexus-standalone
Jan 26 17:34:53.756 DEBG inserted rows into table, table_name: oximeter.measurements_cumulativeu64, n_rows: 147, id: 0b8beb02-0822-4db0-b5b7-279e06493cc9, component: clickhouse-client, collector_id: 427f0ed7-519f-4fc9-bbf6-df18557bcddb, component: oximeter-standalone, component: nexus-standalone
```

Here's the data, again all zeros, since the vCPUs haven't started:

```
shale :) select * from measurements_cumulativeu64 where timeseries_name = 'virtual_machine:vcpu_usage' order by timestamp desc limit 10;

SELECT *
FROM measurements_cumulativeu64
WHERE timeseries_name = 'virtual_machine:vcpu_usage'
ORDER BY timestamp DESC
LIMIT 10

Query id: bc7a19b4-2fe0-45b2-ba25-e68eb2285a6d

┌─timeseries_name────────────┬───────timeseries_key─┬────────────────────start_time─┬─────────────────────timestamp─┬─datum─┐
│ virtual_machine:vcpu_usage │ 15242972249542590882 │ 2024-01-26 17:34:19.086458482 │ 2024-01-26 17:36:14.211485372 │     0 │
│ virtual_machine:vcpu_usage │  2530575590723472384 │ 2024-01-26 17:34:19.086458482 │ 2024-01-26 17:36:14.211485372 │     0 │
│ virtual_machine:vcpu_usage │ 10851326040477630946 │ 2024-01-26 17:34:19.086458482 │ 2024-01-26 17:36:14.211485372 │     0 │
│ virtual_machine:vcpu_usage │  5352823235966609560 │ 2024-01-26 17:34:19.086458482 │ 2024-01-26 17:36:14.211485372 │     0 │
│ virtual_machine:vcpu_usage │ 17023413341277003966 │ 2024-01-26 17:34:19.086458482 │ 2024-01-26 17:36:14.211485372 │     0 │
│ virtual_machine:vcpu_usage │  7027232757898595956 │ 2024-01-26 17:34:19.086458482 │ 2024-01-26 17:36:14.211485372 │     0 │
│ virtual_machine:vcpu_usage │ 10956216263614091129 │ 2024-01-26 17:34:19.086457232 │ 2024-01-26 17:36:14.211482512 │     0 │
│ virtual_machine:vcpu_usage │ 10771128131807065734 │ 2024-01-26 17:34:19.086457232 │ 2024-01-26 17:36:14.211482512 │     0 │
│ virtual_machine:vcpu_usage │ 18110277017948875871 │ 2024-01-26 17:34:19.086457232 │ 2024-01-26 17:36:14.211482512 │     0 │
│ virtual_machine:vcpu_usage │  6763694628924228580 │ 2024-01-26 17:34:19.086457232 │ 2024-01-26 17:36:14.211482512 │     0 │
└────────────────────────────┴──────────────────────┴───────────────────────────────┴───────────────────────────────┴───────┘

10 rows in set. Elapsed: 0.005 sec.

```

After starting the VM with: `./target/release/propolis-cli --server ::1 state run`, we start to see real usage data:

```
shale :) select * from measurements_cumulativeu64 where timeseries_name = 'virtual_machine:vcpu_usage' order by timestamp desc limit 10;

SELECT *
FROM measurements_cumulativeu64
WHERE timeseries_name = 'virtual_machine:vcpu_usage'
ORDER BY timestamp DESC
LIMIT 10

Query id: 168245e1-be28-4cf8-8f7f-eeaabfc85a25

┌─timeseries_name────────────┬───────timeseries_key─┬────────────────────start_time─┬─────────────────────timestamp─┬────────datum─┐
│ virtual_machine:vcpu_usage │ 15242972249542590882 │ 2024-01-26 17:34:19.086458482 │ 2024-01-26 17:37:14.237624945 │ 167155853974 │
│ virtual_machine:vcpu_usage │  2530575590723472384 │ 2024-01-26 17:34:19.086458482 │ 2024-01-26 17:37:14.237624945 │    357842418 │
│ virtual_machine:vcpu_usage │ 10851326040477630946 │ 2024-01-26 17:34:19.086458482 │ 2024-01-26 17:37:14.237624945 │     20702702 │
│ virtual_machine:vcpu_usage │  5352823235966609560 │ 2024-01-26 17:34:19.086458482 │ 2024-01-26 17:37:14.237624945 │   7501165532 │
│ virtual_machine:vcpu_usage │ 17023413341277003966 │ 2024-01-26 17:34:19.086458482 │ 2024-01-26 17:37:14.237624945 │      1437826 │
│ virtual_machine:vcpu_usage │  7027232757898595956 │ 2024-01-26 17:34:19.086458482 │ 2024-01-26 17:37:14.237624945 │        58121 │
│ virtual_machine:vcpu_usage │ 10956216263614091129 │ 2024-01-26 17:34:19.086457232 │ 2024-01-26 17:37:14.237622035 │    439578704 │
│ virtual_machine:vcpu_usage │ 10771128131807065734 │ 2024-01-26 17:34:19.086457232 │ 2024-01-26 17:37:14.237622035 │   7478365356 │
│ virtual_machine:vcpu_usage │ 18110277017948875871 │ 2024-01-26 17:34:19.086457232 │ 2024-01-26 17:37:14.237622035 │       980502 │
│ virtual_machine:vcpu_usage │  6763694628924228580 │ 2024-01-26 17:34:19.086457232 │ 2024-01-26 17:37:14.237622035 │        61471 │
└────────────────────────────┴──────────────────────┴───────────────────────────────┴───────────────────────────────┴──────────────┘

```

bnaecker commented:
I should have made a the data model for the timeseries itself a bit more clear. We're attaching the silo / project / instance UUIDs, as well as the vCPU ID and the cumulative occupancy time in each of the microstates the hypervisor tracks. For example, for this one VM I'm running now, here is the set of fields for all the timeseries generated by this code:

```
┌───────timeseries_key─┬─silo_id──────────────────────────────┬─project_id───────────────────────────┬─instance_id──────────────────────────┬─vcpu_id─┬─state────┐
│ 16664594675826389592 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       0 │ emu_kern │
│ 11393737791688777248 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       0 │ emu_user │
│  3803554770698379086 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       0 │ idle     │
│  3432773714780367716 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       0 │ init     │
│  5618912762095064089 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       0 │ run      │
│  3070243348358802422 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       0 │ sched    │
│  2993119910425702460 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       1 │ emu_kern │
│ 12993794358256992902 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       1 │ emu_user │
│ 15372851520154047752 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       1 │ idle     │
│ 16177795296466568401 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       1 │ init     │
│  5258000312434609370 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       1 │ run      │
│  2512982362218904848 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       1 │ sched    │
│  6795170569018271440 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       2 │ emu_kern │
│ 18110277017948875871 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       2 │ emu_user │
│ 10771128131807065734 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       2 │ idle     │
│  2236272161634754969 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       2 │ init     │
│ 10956216263614091129 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       2 │ run      │
│  6763694628924228580 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       2 │ sched    │
│ 10851326040477630946 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       3 │ emu_kern │
│ 17023413341277003966 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       3 │ emu_user │
│  5352823235966609560 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       3 │ idle     │
│ 15242972249542590882 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       3 │ init     │
│  2530575590723472384 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       3 │ run      │
│  7027232757898595956 │ 2b21e798-7a3a-494a-9984-622bb06a233f │ 2168c102-9ad5-4a14-b4e6-4a20a913148e │ 02d2f58c-43aa-4ebf-a07b-113c0a446716 │       3 │ sched    │
└──────────────────────┴──────────────────────────────────────┴──────────────────────────────────────┴──────────────────────────────────────┴─────────┴──────────┘
```

I'm wondering if we should shove in the silo / project / instance _names_ as well. Those are not strictly required, but it might be convenient for folks to query on those as well. The field data is extremely small on disk, so I'm generally inclined to add anything we feel we'll need.

bnaecker commented:
I went ahead and added names for the silo / project / instance in [f3eb8b3](https://github.com/oxidecomputer/omicron/pull/4900/commits/f3eb8b3151206712a0ee212ca907c1cd332da08e)

pfmooney commented on oximeter/instruments/src/kstat/virtual_machine.rs:
```diff
@@ -0,0 +1,459 @@
+    // - instance: a kernel-assigned integer
+    // - name: vm -> generic VM info, vcpuX -> info for each vCPU
+    //
+    // At this part of the code, we don't have that kstat instance, only the
+    // virtual machine instance's control plane UUID. However, the VM's "name"
+    // is assigned to be that control plane UUID in the hypervisor. See
+    // https://github.com/oxidecomputer/propolis/blob/759bf4a19990404c135e608afbe0d38b70bfa370/bin/propolis-server/src/lib/vm/mod.rs#L420
+    // for the current code which does that.
+    //
+    // That means we need to indicate interest in both the `vm` and `vcpuX`
+    // kstats for any instance, and then filter to the right instance in the
+    // `to_samples()` method below, because interest is defined on each
```
This really feels like something which propolis should be minding itself, rather than pushing that responsibility to oximeter and/or sled-agent.  It's propolis which chooses the naming for the kernel vmm resource, and should probably query the kstats.

Furthermore, I'm not sure how much of the kstat structure we want to turn into a Committed interface in the data.  Do we want to decide on a more constrained model for how vCPUs can spend their time, rather than directly encoding the kstat names (`time_emu_kern`, etc) into the data?

bnaecker replied:
And to be clear about the mapping from the actual kstat run states to those three proposed, I think we'd have:

- `run == time_run + time_emu_kern + time_emu_user`
- `idle == time_init + time_idle`
- `waiting == time_sched`

pfmooney replied:
> Propolis _is_ actually querying the kstats, even though the functionality for converting from that to oximeter `Sample`s lives here.

If propolis is otherwise calling into an abstraction provided by oximeter to do that querying, I'm not sure how much that counts as doing the querying.  Propolis has tight coupling to the kernel vmm interfaces (out of necessity), with explicit logic for handling when that interface undergoes change.

> I do think there's value in the `VirtualMachine` definition living in a public, shared place. I'm trying to avoid someone having a conflicting definition, or even a duplicate definition.

Agreed.

> If that's not desirable, I think it's fine to move the target / metric definitions and the `impl KstatTarget` both to propolis. It does mean that anyone wishing to publish oximeter stats about the VM would need to either depend on Propolis or copy that `Target` implementation. I honestly don't feel that strongly, since neither seems without its own drawbacks.

I think something like that would be the right plan.  I'm not sure where, other than propolis, we would expect VM-related oximeter metrics to come from.

> If you were debugging unexpected behavior in Propolis, would you like to be able to see each of these states historically?

If I'm debugging propolis, then sure, but I'm not sure that detail is worth the trade-off of committing to those states as the interface.

> Maybe three states: _doing work_, _not doing work_, and _not able to do work_ (waiting for some reason)? That would be something like `run`, `idle`, and `waiting`.

Yeah, something like what you described sounds right.  Perhaps we might want to split out time spent in emulation (not distinguishing to the consumer whether it's userland or kernel).

bnaecker replied:
Sweet, that all sounds good to me. To recap:

- Move the `VirtualMachine` and `VcpuUsage` types to Propolis
- Move `impl KstatTarget for VirtualMachine` to Propolis too
- Specify states as: `run`, `emulation`, `idle`, and `blocked`

If that sounds fine, I'll get it moving this afternoon.

bnaecker commented:
Ok, aside for cherry-picking some updates to Progenitor in 2998979, this is ready for another review @pfmooney and @gjcolombo. I've re-run the same tests I described earlier, but against an updated version of [this Propolis branch](https://github.com/oxidecomputer/propolis/tree/vcpu-usage-stats). I've moved the `VirtualMachine` definition to Propolis, and the `impl KstatTarget` for it. I've also added logic for aggregating the states as discussed above, and updated the tests to confirm that works.

bnaecker commented:
Ok, the cherry-picked commit is now in `main`; I've rebased this on that; and I've run the same tests again. Things still look good.

bnaecker commented:
Friendly ping on this @gjcolombo @pfmooney, I'd like to get this merged before things drift too far from `main` again.

bnaecker commented:
I realized Patrick is out this week, and Greg is in a crunch before pat leave. There's also not much here specific to vCPU stats anymore, since the heavy lifting is now in Propolis. I'm going to add a few more reviewers to move things along.

bnaecker commented on sled-agent/src/metrics.rs:
```diff
@@ -258,3 +250,26 @@ impl MetricsManager {
+        //
+        // Safety: We've passed an extra NULL to the gethostname(2) call, so
+        // there must be at least one of these.
+        let chunk = out.split(|x| *x == 0).next().unwrap();
```
That seems reasonable. Another option is querying the hostname length directly, with `sysconf(_SC_HOST_NAME_MAX)` or similar. I was hoping to make this infallible so that we can always produce data, but that might not be realistic.

bnaecker commented on nexus/src/app/instance.rs:
```diff
@@ -1183,6 +1183,28 @@ impl super::Nexus {
+        let (_, db_silo) = self.current_silo_lookup(opctx)?.fetch().await?;
+        let metadata = sled_agent_client::types::InstanceMetadata {
+            silo_id: db_silo.id(),
+            silo_name: db_silo.name().to_string(),
+            project_id: db_project.id(),
+            project_name: db_project.name().to_string(),
```
I think this is a reasonable place to draw the line. It also has some drawbacks. It means Nexus needs to understand more of the query structure, so that it can do things like translate a silo name into a silo ID. If there's some complicated query passed in, that might be a lot of work. That may also be unavoidable. I'm still not sure how auth will intersect with querying, but Nexus might already need to know what components of a query need to be checked for authz. E.g., Nexus would already need to be able to understand the query enough to know that something like `silo_name == "my-silo"` involves some authz checks.

As an alternative, we could imagine including the names in the metrics, and then require that updates to those names be propagated to the relevant leaves. For example, if a user updates a silo name, then we might have a task in Nexus that sends the new name(s) to all the sled-agents hosting instances in that silo. The sled-agents could then forward those to the relevant Propolis servers as well, which could update the instance metadata they publish to oximeter.

That has drawbacks too. The data may include names that do not exist as far as the Nexus API is concerned, which might be confusing or present authz problems of its own. (How do you check that a user is allowed to see metrics from a silo, if that silo no longer has that name? The name-to-ID lookup will fail.) We might be able to just obviate a lot of these by requiring that queries only refer to immutable IDs, rather than updateable names.

I'm not sure the right answer here.

bnaecker replied:
Sounds good, will update soon.

bnaecker replied:
Removed in 5ac187ddd

bnaecker commented:
Ok, all comments addressed in 5ac187ddd, this should be ready for a re-review @smklein

smklein commented on oximeter/producer/src/lib.rs:
```diff
@@ -251,11 +251,19 @@ pub async fn register(
+    client.cpapi_producers_post(&server_info.into()).await.map(|_| ()).map_err(
+        |err| {
+            let retryable = match &err {
+                nexus_client::Error::CommunicationError(..) => true,
+                nexus_client::Error::ErrorResponse(resp) => {
+                    resp.status().is_server_error()
+                }
+                _ => false,
```
I guess it's fine to still stored this in the `RegistrationError` error type, just a little odd that we're still pulling out `retryable` but never using (or ever printing it).

smklein commented on oximeter/instruments/Cargo.toml:
```diff
@@ -5,27 +5,48 @@ edition = "2021"
+tokio = { workspace = true, optional = true }
+thiserror = { workspace = true, optional = true }
+uuid = { workspace = true, optional = true }
+omicron-workspace-hack = { workspace = true, optional = true }
```
Is this actually an optional dep? Don't we want `omicron-workspace-hack` unconditionally?

bnaecker replied:
It's not, I'll fix that thanks.

bnaecker commented:
I can't reply to the other comment for some reason, but that `retryable` field is used in Propolis. I remember now why I did that -- Propolis spins indefinitely trying to register as a metric producer with Nexus. This field is used there to determine whether to backoff and try again, or bail that registration.
